src/kakarot/accounts/library.cairo

```diff
@@ -15,64 +18,311 @@ from starkware.starknet.common.syscalls import (
     }(kakarot_address: felt, _evm_address) {
-        let (is_initialized) = is_initialized_.read();
+        let (is_initialized) = Account_is_initialized.read();
         assert is_initialized = 0;
-        is_initialized_.write(1);
+        Account_is_initialized.write(1);
         Ownable.initializer(kakarot_address);
-        evm_address.write(_evm_address);
+        Account_evm_address.write(_evm_address);
+        Account_kakarot_address.write(kakarot_address);
+
         // Give infinite ETH transfer allowance to Kakarot
         let (native_token_address) = IKakarot.get_native_token(kakarot_address);
         let (infinite) = uint256_not(Uint256(0, 0));
         IERC20.approve(native_token_address, kakarot_address, infinite);
         return ();
     }
 
-    // @return address The EVM address of the contract account
+    // @notice This function is used to upgrade the smart contract account.
+    // @param new_class The new class of the account.
+    func upgrade{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }(new_class: felt) {
+        // Access control check. Only the EOA owner should be able to upgrade its contract.
+        // with `supports_interface`
+        internal.assert_only_self();
+        // TODO: only valid classes should be allowed to be upgraded. Add a validation on the new class interface.
+        assert_not_zero(new_class);
+        replace_class(new_class);
+        Account_implementation.write(new_class);
+        return ();
+    }
+
+    @view
+    func get_implementation{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
+        implementation: felt
+    ) {
+        let (implementation) = Account_implementation.read();
+        return (implementation=implementation);
+    }
+
+    // @return address The EVM address of the account
     func get_evm_address{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}() -> (
         address: felt
     ) {
-        let (address) = evm_address.read();
+        let (address) = Account_evm_address.read();
         return (address=address);
     }
 
+    // @notice This function checks if the account was initialized.
+    // @return is_initialized 1 if the account has been initialized 0 otherwise.
+    func is_initialized{
+        syscall_ptr: felt*,
+        pedersen_ptr: HashBuiltin*,
+        range_check_ptr,
+        bitwise_ptr: BitwiseBuiltin*,
+    }() -> (is_initialized: felt) {
+        let is_initialized: felt = Account_is_initialized.read();
+        return (is_initialized=is_initialized);
+    }
+
+    // EOA functions
+
+    // Constants
+    // @dev see utils/interface_id.py
+    const INTERFACE_ID = 0x68bca1a;
+
+    struct Call {
+        to: felt,
+        selector: felt,
+        calldata_len: felt,
+        calldata: felt*,
+    }
+
+    // Tmp struct introduced while we wait for Cairo to support passing `[Call]` to __execute__
```

Comment:
lol I know that this is the real initial comment, but at this stage, let's not call it `tmp` anymore sa there won't be any change!